### ISSUE TITLE
Azure Storage saga support

### ISSUE THREAD

**Issue by pelin** [rebus-org/Rebus.AzureStorage#8]
Hello! I would like to see Azure Storage saga persistence support for Rebus. It seems the only support right now is for auditing, not for "real" Sagas? I have modified an implementation (https://github.com/rebus-org/Rebus/tree/d0f1ad04ce66612d93a5cb034b8923c15797c136/Rebus.AzureStorage/Sagas)  to support this and also found it to work very well. I have added concurrency support, for example. I'd like to do share this implementation, but I see that this repos is to be deprecated... How come? As I understand it will be difficult to add Saga support in the Azure Blobs repo or in the Azure Queues repos as we need both table and blob support to cover the need.
//Peter

**Comment by mookid8000** [comment]
I removed the old saga persister, because it was not bulletproof, concurrency-wise... IIRC, the problem was when a saga data instance in a blob was updated, its correlation properties had to be updated at the same time, in a way that prevented a message to be correlated with the previous correlation properties and then hitting the new saga state.

Do you have any ideas on how this could be solved?

**Comment by pelin** [comment]
Yes, I do. Although there is one simplification I've made, i.e. no additional correlation properties can be added after the saga once have started. This could be fixed if needed, pretty easy I guess... 
Basically there are two key changes to get it to work, and they are interrelated:
-  the "revision" is added as metadata on the blob, and its always the blob that holds the concurrency responsibility. I manually check the incoming revision on the sagadata against the existing blob from the ISagaStorage.Update-operation.
- the need to update the table storage is removed. Before the revision was written to the table storage. This is removed and now moved to the blob metadata. The table storage indexes are written only once, and as the "secondary indexes" are written first, there will be race condition/conflict/exception if two sagas on the same "primary key" is running in parallell.

Attach the class
[AzureStorageSagaStorage.txt](https://github.com/rebus-org/Rebus.AzureStorage/files/3569252/AzureStorageSagaStorage.txt)

**Comment by mookid8000** [comment]
ok that actually sounds like it could work... I'm also slowly considering whether _updating_ correlation properties should be possible at all.... it leads to all kinds of annoying concurrency problems, a whole class of problems that would vanish if only they were immutable.

People usually use them to "abandon" deferred messages that the saga would otherwise have received... but there's other ways to do that.

**Comment by pelin** [comment]
Yes, I think its a good idea to make the correlation properties immutable. One compromise would be to only add new correlation properties, i.e. never update or delete existing records. But I guess its more easy to understand for users to say its completely immutable after creation of the saga...

It would be great if you will consider to send the code into production/master, and revive the Rebus.AzureStorage repo.

**Comment by dannythunder** [comment]
Working on a PoC with Rebus sagas, glad to see you're on to this @pelin :-D Will talk to you tomorrow on how to implement this and possible problems.

**Comment by mookid8000** [comment]
Hi @pelin and @dannythunder , any progress on this?

**Comment by dannythunder** [comment]
Sorry, nope. I think it was put on hold.
